lib/streamlit/connections/snowflake_connection.py

```diff
@@ -0,0 +1,260 @@
+
+        # Otherwise, attempt to create a new connection from whatever credentials we
+        # have available.
+
+        st_secrets = self._secrets.to_dict()
+        if len(st_secrets):
+            # TODO(vdonato): Double-check that we want to merge any secrets we find with
+
+            # the kwargs passed to `st.connection`.
+            conn_kwargs = {**st_secrets, **kwargs}
+            _validate_connection_params(conn_kwargs)
+            return snowflake.connector.connect(**conn_kwargs)
+
+        if hasattr(snowflake.connector.connection, "CONFIG_MANAGER"):
+            # TODO(vdonato): Double-check that we want to map the "snowflake" connection
+            # name in a call to `st.connection` to the "default" connection in the
+            # Snowflake config file. This is aligned with the default behavior for the
+            # underlying python connector and seems more sane than having a "snowflake"
+            # connection in a snowflake-specific config file.
+            connection_name = (
+                "default"
+                if self._connection_name == "snowflake"
+                else self._connection_name
+            )
+            return snowflake.connector.connect(
+                connection_name=connection_name,
+                **kwargs,
+            )
+
+        # If we have a version of the Snowflake Python Connector installed that predates
+        # the ConfigManager, look for credentials in  ~/.snowsql/config as a last
+        # resort.
+        snowsql_config = load_from_snowsql_config_file(self._connection_name)
+        conn_kwargs = {**snowsql_config, **kwargs}
+        _validate_connection_params(conn_kwargs)
+
+        return snowflake.connector.connect(**conn_kwargs)
+
+    def query(
+        self,
+        sql: str,
+        *,  # keyword-only arguments:
+        ttl: float | int | timedelta | None = None,
+        show_spinner: bool | str = "Running `snowflake.query(...)`.",
+        params=None,
+    ) -> pd.DataFrame:
+        """Run a read-only SQL query.
+
+        This method implements both query result caching (with caching behavior
+        identical to that of using ``@st.cache_data``) as well as simple error handling/retries.
+
+        .. note::
+            Queries that are run without a specified ttl are cached indefinitely.
+
+        Parameters
+        ----------
+        sql : str
+            The read-only SQL query to execute.
+        ttl : float, int, timedelta or None
+            The maximum number of seconds to keep results in the cache, or
+            None if cached results should not expire. The default is None.
+        show_spinner : boolean or string
+            Enable the spinner. The default is to show a spinner when there is a
+            "cache miss" and the cached resource is being created. If a string, the value
+            of the show_spinner param will be used for the spinner text.
+        params : list, tuple, dict or None
```

Comment:
As noted in Slack, when running outside of SiS I think we need to enable the qmark binding style by default, so code works across both. This is also consistent with the snowpark params support ([cite](https://docs.snowflake.com/en/developer-guide/snowpark/reference/python/latest/api/snowflake.snowpark.Session.sql))

Reply:
👍